File: Seattle_stats/data_processing.py
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
# I am going to analyze some of the data available through <data.seattle.gov>.

# Data currently being considered:
#
#  1.City of Seattle, Department of Information Technology,
#      Seattle Police Department
#    "Seattle Police Department Police Report Incident"
#    Downloaded 2018-09-16 from
#    https://data.seattle.gov/api/views/7ais-f98f/rows.csv?accessType=DOWNLOAD
#  
#  2.City of Seattle, Department of Information Technology, Seattle Police
#    Department
#    "Seattle Police Department 911 Incident Response"
#    Downloaded 2018-09-16 from
#    https://data.seattle.gov/api/views/3k2p-39jp/rows.csv?accessType=DOWNLOAD
#  
#  3.Seattle Police Department (http://www.seattle.gov/police/)
#    "Call Data"
#    Downloaded 2018-09-16 from
#    https://data.seattle.gov/api/views/33kz-ixgy/rows.csv?accessType=DOWNLOAD
#  
#  4.Seattle Police Department (http://www.seattle.gov/police/)
#    "Crime Data"
#    Downloaded 2018-09-16 from
#    https://data.seattle.gov/api/views/4fs7-3vj5/rows.csv?accessType=DOWNLOAD
#  
#  5.City of Seattle, Department of Information Technology
#      (http://seattle.gov/police/crime/default.htm)
#    "Seattle Crime Stats by Police Precinct 2008-Present"
#    Downloaded 2018-09-16 from
#    https://data.seattle.gov/api/views/3xqu-vnum/rows.csv?accessType=DOWNLOAD
#  
#  6.Department of Transportation
#    "Road Weather Information Stations"
#    Downloaded 2018-09-16 from
#    https://data.seattle.gov/api/views/egc4-d24i/rows.csv?accessType=DOWNLOAD

# Libraries importing:
#%%= = = = = = = = = = = = = =[ IMPORT LIBRARIES ]= = = = = = = = = = = = = = =
  import sklearn as skl
  import numpy as np
  import scipy as sp
  import matplotlib as mpl
  import pandas as pd
  import datetime as dt
  from datetime import datetime as dt_dt
#%#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#%%= = = = = = = = = = = = = = =[ PARAMETERS ]= = = = = = = = = = = = = = = = =
subset_size=10000
data_dir='./Desktop/career/0_work_samples/Seattle_stats/data/'
#%%= = = = = = = = = = = =[ READ IN CSV AS PANDAS DATA TABLE ]= = = = = = = = =
#Seattle_Police_Department_Police_Report_Incident=pd.read_csv(data_dir+
#    'Seattle_Police_Department_Police_Report_Incident.csv',
#    dtype={
#        
#  'RMS CDW ID':str,
#  'General Offense Number':str,
#  'Offense Code':str,#np.int32,
#  'Offense Code Extension':np.float64,
#  'Offense Type':str,
#  'Summary Offense Code':str,#np.int32,
#  'Summarized Offense Description':str,
#  'Date Reported':str,#dt.datetime,
#  'Occurred Date or Date Range Start':str,#:dt.datetime,
#  'Occurred Date Range End':str,#datetime,
#  'Hundred Block Location':str,
#  'District/Sector':str,
#  'Zone/Beat':str,
#  'Census Tract 2000':np.float64,
#  'Longitude':np.float64,
#  'Latitude':np.float64,
#  'Location':str,
#  'Month':np.int32,
#  'Year':np.int32      
#        
#        }
#    
##    ,nrows=100
#    ,infer_datetime_format=True
#    
#    
#    )
#
##%%
#
#Crime_Data=pd.read_csv(data_dir+'Crime_Data.csv',
#  dtype={
#      'Occurred Time':str,
#      'Reported Time':str
#          },  
#  infer_datetime_format=True
#  )
#
##%%
#Seattle_Police_Department_Police_Report_Incident['Year'][1:100]
##%%
#Seattle_Police_Department_Police_Report_Incident.dtypes
#
##%%
#Seattle_Police_Department_911_Incident_Response=pd.read_csv(data_dir+
#    'Seattle_Police_Department_911_Incident_Response.csv').astype(dtype={'Event Clearance Date':dt.datetime})
##%%
#Seattle_Police_Department_911_Incident_Response.dtypes  
#
#
#
#
#
##%%
#Incident_Response_sub=Seattle_Police_Department_911_Incident_Response[1:100].astype(dtype={'Event Clearance Date':dt.datetime})
##%%
#Incident_Response_sub.dtypes
##%%
#Incident_Response_sub.loc[:,'Event Clearance Date']
#
##%%
#Incident_Response_sub_2=Incident_Response_sub.assign(m=pd.to_datetime(Incident_Response_sub['Event Clearance Date']))
##%%
#Incident_Response_sub_2.columns
#Incident_Response_sub_2.dtypes
##%%
#
#dt_dt.time(Incident_Response_sub_2.m)
#
#Incident_Response_sub_3=Incident_Response_sub_2.assign(d=day(m))
#
##%%
#Seattle_Police_Department_911_Incident_Response.columns
##%%
#Seattle_Police_Department_911_Incident_Response.astype(dtype={'Event Clearance Date':dt.datetime})
#
#
##%%
#dt_dt.strptime(
#    Seattle_Police_Department_911_Incident_Response.loc[:,'Event Clearance Date'].iloc[1:10]
#    ,'%b %d %Y %I:%M:%S:000%p'
#    )
#
##%%
#dt_dt.strptime(calls2.loc[:,'ta'].iloc[i],'%b %d %Y %I:%M:%S:000%p')
#Seattle_Police_Department_911_Incident_Response.loc[:,'Event Clearance Date'][1:10]
#%%
#Call_Data=pd.read_csv(data_dir+'Call_Data.csv')
#%%
Crime_Data=pd.read_csv(data_dir+'Crime_Data.csv',
 dtype={
     'Report Number':str,
     'Occurred Date':str,
     'Occurred Time':str,
     'Reported Date':str,
     'Reported Time':str,
     'Crime Subcategory':str,
     'Primary Offense Description':str,
     'Precinct':str,
     'Sector':str,
     'Beat':str,
     'Neighborhood':str
     }
 )
#%% = = = = = = = = = = = = = = =[ COLUMNS ]= = = = = = = = = = = = = = = = = =
#print(Call_Data.columns);
#  'CAD Event Number'
#  'Event Clearance Description'
#  'Call Type'
#  'Priority'
#  'Initial Call Type'
#  'Final Call Type'
#  'Original Time Queued'
#  'Arrived Time'
#  'Precinct'
#  'Sector'
#  'Beat'

print(Crime_Data.columns);
#  'Report Number'
#  'Occurred Date'
#  'Occurred Time'
#  'Reported Date'
#  'Reported Time'
#  'Crime Subcategory'
#  'Primary Offense Description'
#  'Precinct'
#  'Sector'
#  'Beat'
#  'Neighborhood'
#%% = = = = = = = = = = = = = = =[ ROW COUNTS ]= = = = = = = = = = = = = = = = 
print('len(Call_Data):',len(Call_Data),'\n',
      'len(Crime_Data):',len(Crime_Data),'\n',sep='');
#%% = = = = = = = = = = = = = = =[ SUBSETS ]= = = = = = = = = = = = = = = = = =
#Adding [count] for safe/simple counting.
subset_size=len(Crime_Data);
Call_Data_sub=Call_Data[0:subset_size].assign(count=1);
Crime_Data_sub=Crime_Data[0:subset_size].assign(count=1);
#%% = = = = = = = = = = = = = = =[ TIDYING FORMATS ]= = = = = = = = = = = = = =
Crime_Data_sub_2=Crime_Data[0:subset_size].assign(
    Occurred_Date=pd.to_datetime(Crime_Data_sub['Occurred Date'],format='%m/%d/%Y'),
    Occurred_Time=pd.to_datetime(Crime_Data_sub['Occurred Time'].str.pad(width=4,side='left',fillchar='0'),format='%H%M'),
    Reported_Date=pd.to_datetime(Crime_Data_sub['Reported Date'],format='%m/%d/%Y'),
    Reported_Time=pd.to_datetime(Crime_Data_sub['Reported Time'].str.pad(width=4,side='left',fillchar='0'),format='%H%M')
)
#%% = = = = = = = = = = = = = = =[ MAKE ROOM FOR DATETIMES ]= = = = = = = = = = 
Crime_Data_sub_3=Crime_Data_sub_2.assign(
    Occurred_Datetime='',
    Reported_Datetime='')                
#%% = = = = = = = = = = = = = = =[ ADD DATES AND TIMES ]= = = = = = = = = = = = 
for i in range(0,len(Crime_Data_sub_2)):
  if np.mod(i,1000)==0:
    print(i)
  Crime_Data_sub_3.loc[:,'Occurred_Datetime'].iat[i]=dt_dt.combine(Crime_Data_sub_2.loc[:,'Occurred_Date'].iat[i].date(),
                  Crime_Data_sub_2.loc[:,'Occurred_Time'].iat[i].time())    
  Crime_Data_sub_3.loc[:,'Reported_Datetime'].iat[i]=dt_dt.combine(Crime_Data_sub_2.loc[:,'Reported_Date'].iat[i].date(),
                  Crime_Data_sub_2.loc[:,'Reported_Time'].iat[i].time())    
#%%

##%%    
#Crime_Data_sub_2.columns
#Crime_Data_sub_2.dtypes
#Crime_Data_sub.dtypes
#%%
#a=pd.DataFrame(Report_Incident_sub.groupby(['RMS CDW ID'])['count'].sum())
#a[a['count']>1]
#Report_Incident_sub[pd.isna(Report_Incident_sub['RMS CDW ID'])]
#df1=pd.merge(
#    Call_Data_sub,
#    Incident_Response_sub,
#    how='left',
#    on=['CAD Event Number','CAD Event Number'],
#    suffixes=('__cd','__911'),
#    validate='many_to_one')
#%%= = = = = = = = = = =[ RENAME, TAKE SUBSET FOR DEV ]= = = = = = = = = = = =
#call_data_sub=Call_Data[1:100].rename(index=str, columns={
#    'CAD Event Number':'CAD_event_number',        
#    #- - - - - - - - -[ WHERE ]- - - - - - - - - - - - - - - - - - - - - - - -
#    'Precinct':'precinct',
#    'Sector':'sector',
#    'Beat':'beat',        
#    #- - - - - - - - -[ WHEN ]- - - - - - - - - - - - - - - - - - - - - - - - -
#    'Original Time Queued':'tq',
#    'Arrived Time':'ta',                
#    #- - - - - - - - -[ HOW ]- - - - - - - - - - - - - - - - - - - - - - - - -
#    'Call Type':'channel',        
#    'Priority':'priority',        
#    #- - - - - - - - -[ WHAT ]- - - - - - - - - - - - - - - - - - - - - - - - -
#    'Initial Call Type':'description_i',
#    'Final Call Type':'description_f',
#    'Event Clearance Description':'resolution'
#  })[0:subset_size].assign(count=1);
#print(calls.columns)
#print(len(calls))
#%%= = = = = = = = = = = = =[ REFORMAT TIME FIELDS ]= = = = = = = = = = = = = =
#calls2=calls.assign(
#    tq2=dt.datetime(1700,1,1),
#    ta2=dt.datetime(1700,1,1)
#  );
#for i in range(0,len(calls2)):
#  if np.mod(i+1,1000)==0:
#    print(i+1)
#  calls2.loc[:,'ta2'].iloc[i]= \
#    dt_dt.strptime(calls2.loc[:,'ta'].iloc[i],'%b %d %Y %I:%M:%S:000%p')
#  calls2.loc[:,'tq2'].iloc[i]= \
#    dt_dt.strptime(calls2.loc[:,'tq'].iloc[i],'%m/%d/%Y %I:%M:%S %p')
#print(calls2.loc[:,['ta2','tq2']].iloc[1:1000])    
#%#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

File: Seattle_stats/data_processing.R
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
# I am going to analyze some of the data available through <data.seattle.gov>.

# Data used by this script:
#  
#  Seattle Police Department (http://www.seattle.gov/police/)
#    "Crime Data"
#    Downloaded 2018-09-16 from
#    https://data.seattle.gov/api/views/4fs7-3vj5/rows.csv?accessType=DOWNLOAD

# Libraries loaded:
library(dplyr)
library(data.table)
library(ggplot2)
library(stats)
library(stringr)
library(lubridate)
library(tidyr)

#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


# = = = = = = = = = = = = =[ READ IN DATA ]= = = = = = = = = = = = = = = = = =
dir_path=paste0('C:\\Users\\Emily\\Desktop\\career\\0_work_samples\\',
 'Seattle_stats\\')
Crime_Data_o<-read.csv(paste0(dir_path,'data\\Crime_Data.csv'),sep=',')

#= = = = = = = = = = = = = =[ RENAME, TAKE SUBSET ]= = = = = = = = = = = = = =
# n=1000
n=nrow(Crime_Data_o)
crime_data<-data.frame(
  report_number=Crime_Data_o$Report.Number[1:n],
  occurred_date=Crime_Data_o$Occurred.Date[1:n],
  occurred_time=Crime_Data_o$Occurred.Time[1:n],
  reported_date=Crime_Data_o$Reported.Date[1:n],
  reported_time=Crime_Data_o$Reported.Time[1:n],
  crime_subcategory=Crime_Data_o$Crime.Subcategory[1:n],
  primary_offense_description=Crime_Data_o$Primary.Offense.Description[1:n],
  precinct=Crime_Data_o$Precinct[1:n],
  sector=Crime_Data_o$Sector[1:n],
  beat=Crime_Data_o$Beat[1:n],
  neighborhood=Crime_Data_o$Neighborhood[1:n]
)

#= = = = = = = = =[ REFORMAT DATES, TAKE DIFF, AGG CATEGORIES ]= = = = = = = =
df<-crime_data%>%
  mutate(
    occurred_datetime=
      as.POSIXct(strptime(paste0(
        occurred_date,
        substr(paste0(10000+occurred_time),2,5)
      ),format='%m/%d/%Y%H%M')),
    reported_datetime=
      as.POSIXct(strptime(paste0(
        reported_date,
        substr(paste0(10000+reported_time),2,5)
      ),format='%m/%d/%Y%H%M')),
    dt=
      as.numeric(
        difftime(
        reported_datetime,
        occurred_datetime,
        units='d')),
    crime_subcategory_2=
      ifelse(crime_subcategory%like%'AGGRAVATED ASSAULT','agg.assault',
      ifelse(crime_subcategory%like%'ARSON','arson',
      ifelse(crime_subcategory%like%'BURGLARY','burglary',
      ifelse(crime_subcategory%like%'DUI','DUI',
      ifelse(crime_subcategory%like%'HOMICIDE','homicide',
      ifelse(crime_subcategory%like%'NARCOTIC','narcotic',
      ifelse(crime_subcategory%like%'OTHER','(other)',
      ifelse(crime_subcategory%like%'PROSTITUTION','prost.',
      ifelse(crime_subcategory%like%'RAPE','rape',
      ifelse(crime_subcategory%like%'ROBBERY','robbery',
      ifelse(crime_subcategory%like%'THEFT','theft','other'
     ))))))))))))


#=================[ dt DISTRIBUSTIONS BY CRIME TYPE BEGINS ]==================

#- - - - - - - - - - - - -[ X BOUNDS, log_10*(dt) ]- - - - - - - - - - - - - -
#Take log10 of dt. If dt is 0, count as log10(min(dt>0))-bin_width.
#If dt is <0, count as na.
bin_width=0.5
zerosub<-df%>%filter(dt>0)%>%summarize(zerosub=
  floor(log10(min(dt)))-bin_width)%>%as.numeric()
log_min<-df%>%filter(dt>0)%>%summarize(log_min=
  floor(log10(min(dt))))%>%as.numeric()
log_max<-df%>%filter(dt>0)%>%summarize(log_max=
  ceiling(log10(max(dt))))%>%as.numeric()
df<-df%>%mutate(log_dt=ifelse(dt>0,log10(dt),ifelse(is.na(dt),NA,zerosub)))


#----------------------------[ FACET GRIDS BEGINS ]---------------------------
#- - - - - - - - -[ HOMICIDE/AGGRAVATED ASSAULT/RAPE/ROBBERY ]- - - - - - - -
dists_1<-ggplot(df%>%filter(is.element(crime_subcategory_2,c(
  'homicide',
  'agg. assault',
  'rape',
  'robbery'
  ))))+
stat_bin(aes(x=log_dt,y=stat(density)),
  color="black",
  fill="light green",
  breaks=c(zerosub,seq(log_min,log_max,bin_width))
  )+
scale_x_continuous(
  breaks=c(zerosub+(bin_width/2),
           as.integer(seq(log_min,log_max,bin_width))+(bin_width/2)),
  labels=c('-Inf',as.integer(seq(log_min,log_max,bin_width)))
  )+
scale_y_discrete()+
labs(title="time from crime to reporting",
  x=expression('log'[10]*'(days)'))+
facet_grid(crime_subcategory_2~.)

jpeg(paste0(dir_path,'output\\dists_1.jpg'))
print(dists_1)
dev.off()
print(dists_1)

#- - - - - - - - - - - - -[ THEFT/ARSON/BURGLARY ]- - - - - - - - - - - - - -
dists_2<-ggplot(df%>%filter(is.element(crime_subcategory_2,c(
  'theft',
  'arson',
  'burglary'
))))+
  stat_bin(aes(x=log_dt,y=stat(density)),
           color="black",
           fill="light green",
           breaks=c(zerosub,seq(log_min,log_max,bin_width))
  )+
  scale_x_continuous(
    breaks=c(zerosub+(bin_width/2),
             as.integer(seq(log_min,log_max,bin_width))+(bin_width/2)),
    labels=c('-Inf',as.integer(seq(log_min,log_max,bin_width)))
  )+
  scale_y_discrete()+
  labs(title="time from crime to reporting",
    x=expression('log'[10]*'(days)'))+
  facet_grid(crime_subcategory_2~.)

jpeg(paste0(dir_path,'output\\dists_2.jpg'))
print(dists_2)
dev.off()
print(dists_2)

#- - - - - - - - - -[ DUI/NARCOTIC/PROSTITUTION/OTHER ]- - - - - - - - - - - -
dists_3<-ggplot(df%>%filter(is.element(crime_subcategory_2,c(
  'DUI',
  'narcotic',
  'prost.',
  '(other)'
))))+
  stat_bin(aes(x=log_dt,y=stat(density)),
           color="black",
           fill="light green",
           breaks=c(zerosub,seq(log_min,log_max,bin_width))
  )+
  scale_x_continuous(
    breaks=c(zerosub+(bin_width/2),
             as.integer(seq(log_min,log_max,bin_width))+(bin_width/2)),
    labels=c('-Inf',as.integer(seq(log_min,log_max,bin_width)))
  )+
  scale_y_discrete()+
  labs(title="time from crime to reporting",
    x=expression('log'[10]*'(days)'))+
  facet_grid(crime_subcategory_2~.)

jpeg(paste0(dir_path,'output\\dists_3.jpg'))
print(dists_3)
dev.off()
print(dists_3)

#==================[ dt DISTRIBUSTIONS BY CRIME TYPE ENDS ]===================


# = = = = = = = =[ AVERAGE DAYS BETWEEN OCCURANCE AND REPORTING ]= = = = = = =
df_sum<-df%>%
  group_by(crime_subcategory)%>%
  summarize(
    count=n(),
    dt_mean=mean(dt,na.rm=TRUE),
    dt_median=median(dt,na.rm=TRUE)
  )
View(df_sum)

df_sum<-df%>%
  group_by(crime_subcategory,year=year(reported_datetime))%>%
  summarize(
    count=n(),
    dt_mean=mean(dt,na.rm=TRUE),
    dt_median=median(dt,na.rm=TRUE)
    )%>%arrange(year,.by_group=TRUE)
View(df_sum)

#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%